clients/arweave_ipfs_s3.go

```diff
@@ -2,52 +2,138 @@ package clients
+	var err error
+	var gateways []*url.URL
+	dStorageURL, err := url.Parse(u)
 	if err != nil {
-		return fmt.Errorf("error fetching Arweave or IPFS URL: %s", err)
+		return nil, fmt.Errorf("invalid URL: %w", err)
 	}
-	defer resp.Body.Close()
 
-	err = UploadToOSURL(s3URL, "", resp.Body, MAX_COPY_DURATION)
+	var resourceID string
+
+	if dStorageURL.Scheme == SCHEME_ARWEAVE {
+		gateways = config.ImportArweaveGatewayURLs
+		resourceID = dStorageURL.Host
+	} else if dStorageURL.Scheme == SCHEME_IPFS {
+		gateways = config.ImportIPFSGatewayURLs
+		resourceID = dStorageURL.Host
+	} else {
+		var gateway, dStorageType string
+		resourceID, gateway, dStorageType = parseDStorageGatewayURL(dStorageURL)
+		if dStorageType == "" {
+			return nil, fmt.Errorf("unsupported dStorage resource %s", dStorageURL.Scheme)
+		}
+
+		gatewayURL, err := url.Parse(gateway)
+		if err != nil {
+			return nil, fmt.Errorf("invalid URL: %w", err)
+		}
+
+		gateways = append(gateways, gatewayURL)
+		if dStorageType == "ar" {
+			gateways = append(gateways, config.ImportArweaveGatewayURLs...)
+		} else {
+			gateways = append(gateways, config.ImportIPFSGatewayURLs...)
+		}
+	}
+
+	var opContent io.ReadCloser
+	downloadOperation := func() error {
+		for _, gateway := range gateways {
+			opContent = downloadDStorageResourceFromSingleGateway(gateway, resourceID, requestID)
+			if opContent != nil {
+				return nil
+			}
+		}
+
+		return fmt.Errorf("failed to fetch %s from any of the gateways", u)
+	}
+
+	retryStrategy := backoff.NewConstantBackOff(1 * time.Second)
+	err = backoff.Retry(downloadOperation, backoff.WithMaxRetries(retryStrategy, 2))
 	if err != nil {
-		return fmt.Errorf("failed to copy Arweave or IPFS URL to S3: %s", err)
+		return nil, err
+	} else {
+		return opContent, nil
 	}
+}
 
-	return nil
+func downloadDStorageResourceFromSingleGateway(gateway *url.URL, resourceId, requestID string) io.ReadCloser {
+	fullURL := gateway.JoinPath(resourceId).String()
+	resp, err := http.DefaultClient.Get(fullURL)
+
+	if err != nil {
+		log.LogError(requestID, "failed to fetch content from gateway", err, "url", fullURL)
+		return nil
+	}
+
+	if resp.StatusCode >= 300 {
+		resp.Body.Close()
+		log.Log(requestID, "unexpected response from gateway", "status_code", resp.StatusCode, "url", fullURL)
+		return nil
+	}
+
+	return resp.Body
 }
 
-func IsArweaveOrIPFSURL(arweaveOrIPFSURL string) bool {
-	u, err := url.Parse(arweaveOrIPFSURL)
+func IsDStorageResource(dStorage string) bool {
+	u, err := url.Parse(dStorage)
 	if err != nil {
 		return false
 	}
 
-	return strings.Contains(u.Host, "arweave") || strings.Contains(u.Host, "w3s.link") || strings.Contains(u.Path, "ipfs")
+	if u.Scheme == SCHEME_ARWEAVE || u.Scheme == SCHEME_IPFS {
+		return true
+	}
+	_, _, dStorageType := parseDStorageGatewayURL(u)
+
+	return dStorageType != ""
+}
+
+func parseDStorageGatewayURL(u *url.URL) (string, string, string) {
+	if strings.Contains(u.Host, "arweave") {
```

Comment:
Not a regression, but IMO this should be an exact check as there is only 1 known arweave gateway that this was trying to match AFAIK.
```suggestion
	if u.Host == "arweave.net" {
```